### ISSUE TITLE
[FEATURE] Support `jsonschema` 4.18.0+

### ISSUE THREAD

**Issue by allanlewis** [schemathesis/schemathesis#1755]
**Is your feature request related to a problem? Please describe.**

As of v4.18.0, `jsonschema` emits the following warnings:

```
DeprecationWarning: jsonschema.exceptions.RefResolutionError is deprecated as of version 4.18.0. If you wish to catch potential reference resolution errors, directly catch referencing.exceptions.Unresolvable.
DeprecationWarning: jsonschema.RefResolver is deprecated as of v4.18.0, in favor of the https://github.com/python-jsonschema/referencing library, which provides more compliant referencing behavior as well as more flexible APIs for customization. A future release will remove RefResolver. Please file a feature request (on referencing) if you are missing an API for the kind of customization you need.
```

**Describe the solution you'd like**

It would be good if this package could address these warning before these deprecations turn into removals.

**Describe alternatives you've considered**

`jsonschema` could be pinned to older versions, but we would need to support 4.18.x at some point.

**Additional context**

I have raised a similar issue in `hypothesis-jsonschema`: python-jsonschema/hypothesis-jsonschema#102

**Comment by allanlewis** [comment]
I see that python-jsonschema/hypothesis-jsonschema#102 appears to have been resolved in 0.23, but this package explicitly limits `hypothesis-jsonschema` to `<0.23` 😕

**Comment by Stranger6667** [comment]
Originally I wanted to update hypothesis-jsonschema and provide an api to control the generated characters, but later postponed it. Probably, the update itself should be compatible, so maybe a patch version would work

**Comment by Stranger6667** [comment]
I am going to postpone it a bit more - as `jsonschema` moves towards the `referencing` library, it requires quite more work to properly integrate it. Additionally, I noticed that `hypothesis-jsonschema` uses some private `jsonschema` APIs (e.g. `_RefResolutionError` that is deprecated), so it should be updated too.

But the next release will support `hypothesis-jsonschema==0.23`, so there are much fewer deprecation warnings.

In any event, I made some progress towards integrating `referencing` and will focus on it after this release.

**Comment by allanlewis** [comment]
Is this resolved now? I've lost track...

**Comment by Stranger6667** [comment]
No, my plan is to finish v4 (soon) and port everything to jsonschema_rs instead of (mostly due to ownership & performance reasons)

**Comment by Stranger6667** [comment]
Btw, for the time being, I am going to silence the warnings from jsonschema as the deprecated parts are going to be removed in jsonschema 5, but Schemathesis has it pinned to version 4

**Comment by Stranger6667** [comment]
Going to close this one. `jsonschema` is pinned to v4 and all the deprecated imports are handled explicitly and deprecation warnings are suppressed as there is no plan to use the next major version of `jsonschema`. Migration to `jsonschema_rs` will be done separately and it is quite a large task on its own
